clang/lib/Sema/AnalysisBasedWarnings.cpp

```diff
@@ -2397,6 +2397,1262 @@ class UnsafeBufferUsageReporter : public UnsafeBufferUsageHandler {
 };
 } // namespace
 
+// =============================================================================
+
+namespace FXAnalysis {
+
+enum class DiagnosticID : uint8_t {
+  None = 0, // sentinel for an empty Diagnostic
+  Throws,
+  Catches,
+  CallsObjC,
+  AllocatesMemory,
+  HasStaticLocal,
+  AccessesThreadLocal,
+
+  // These only apply to callees, where the analysis stops at the Decl
+  DeclDisallowsInference,
+
+  CallsDeclWithoutEffect,
+  CallsExprWithoutEffect,
+};
+
+// Holds an effect diagnosis, potentially for the entire duration of the
+// analysis phase, in order to refer to it when explaining why a caller has been
+// made unsafe by a callee.
+struct Diagnostic {
```

Comment:
This class is serving as an abstraction to unify the warning when a `nonblocking` (etc) function blocks, vs. the note when a function's caller is attempting to infer it as `nonblocking` and fails because the function blocks.

There are many (6?) of these diagnostics that come in pairs, a warning and a note. This `Diagnostic` class represents the exact unsafe construct and has no opinion about whether it needs to be reported as a warning or a note; there's one function that converts it into a regular diagnostic given the context.

Reply:
I see, yeah. I wonder if there might be a better name for this that makes this a bit clearer—I’d at least mention in a comment that this class is less a diagnostic and more a (partial) result of the analysis which is later used to issue diagnostics.

Reply:
I agree that another name would be better. In my next push it is a "Violation" (like a parking violation), unless someone has a better idea. And there's a comment about how it abstracts the contexts.

Reply:
Yeah, something like that works better imo (I’m mainly against the term ‘diagnostic’ in particular because that already has such an established meaning in Clang...)